bentoml/utils/docker_utils.py

```diff
@@ -0,0 +1,179 @@
+import click
+import sys
+import os
+import json
+import re
+import psutil
+from bentoml.saved_bundle import load_bento_service_metadata
+from bentoml.cli.utils import Spinner, echo_docker_api_result
+from bentoml.exceptions import BentoMLException, CLIException
+from bentoml.utils import resolve_bundle_path
+from bentoml.cli.click_utils import (
+    CLI_COLOR_WARNING,
+    CLI_COLOR_SUCCESS,
+    _echo,
+    BentoMLCommandGroup,
+    conditional_argument,
+)
+
+
+def to_valid_docker_image_name(name):
+    # https://docs.docker.com/engine/reference/commandline/tag/#extended-description
+    return name.lower().strip("._-")
+
+
+def to_valid_docker_image_version(version):
+    # https://docs.docker.com/engine/reference/commandline/tag/#extended-description
+    return version.encode("ascii", errors="ignore").decode().lstrip(".-")[:128]
+
+
+def validate_tag(ctx, param, tag):  # pylint: disable=unused-argument
+    if tag is None:
+        return tag
+
+    if ":" in tag:
+        name, version = tag.split(":")[:2]
+    else:
+        name, version = tag, None
+
+    valid_name_pattern = re.compile(
+        r"""
+        ^(
+        [a-z0-9]+      # alphanumeric
+        (.|_{1,2}|-+)? # seperators
+        )*$
+        """,
+        re.VERBOSE,
+    )
+    valid_version_pattern = re.compile(
+        r"""
+        ^
+        [a-zA-Z0-9] # cant start with .-
+        [ -~]{,127} # ascii match rest, cap at 128
+        $
+        """,
+        re.VERBOSE,
+    )
+
+    if not valid_name_pattern.match(name):
+        raise click.BadParameter(
+            f"Provided Docker Image tag {tag} is invalid. "
+            "Name components may contain lowercase letters, digits "
+            "and separators. A separator is defined as a period, "
+            "one or two underscores, or one or more dashes.",
+            ctx=ctx,
+            param=param,
+        )
+    if version and not valid_version_pattern.match(version):
+        raise click.BadParameter(
+            f"Provided Docker Image tag {tag} is invalid. "
+            "A tag name must be valid ASCII and may contain "
+            "lowercase and uppercase letters, digits, underscores, "
+            "periods and dashes. A tag name may not start with a period "
+            "or a dash and may contain a maximum of 128 characters.",
+            ctx=ctx,
+            param=param,
+        )
+    return tag
+
+
+def containerize_bento_service(
+    bento, push, tag, build_arg, username, password, pip_installed_bundle_path
```

Comment:
Do we use `pip_installed_bundle_path` anywhere?

Since bento is only required parameter, let's make rest of them optional

Reply:
'containerize' command seems to use it

Reply:
You are correct.  The CLI command is using it, so we can provide cli commands like:
```
$ pip install iris_classifier_bento_bundle
$ IrisClassifier containerize
```
When you are inisde Yatai deployment operator, you have access to the bento bundle path via the Bento protobuf, you won't need to figure out the bundle path.

For function like this, we want it single purpose. And since containerize function is going to be move to yatai services, let's not refactor right now.  We will do a big refactoring then.

If you want an example of building docker image and push, you can check out the azure-functions deployment operator.  https://github.com/bentoml/BentoML/blob/3f942f273c7bd4d909fb4ea1ee6cc1f3c75599be/bentoml/yatai/deployment/azure_functions/operator.py#L202

Reply:
I thought docker building and pushing will be used by many services,so made it general purpose.
But if you want to refactor later thats ok too.I will keep them separate for deployment and cli.